labelbox/schema/dataset.py

```diff
@@ -749,3 +765,94 @@ def _export(
         task_id = res["taskId"]
         return Task.get_task(self.client, task_id)
+
+    def upsert_data_rows(self, specs: List[DataRowSpec]) -> "Task":
+        """
+        Upserts data rows in this dataset.
+
+        >>>     task = dataset.upsert_data_rows([
+        >>>         # create new data row
+        >>>         DataRowSpec(
+        >>>             row_data="http://my_site.com/photos/img_01.jpg",
+        >>>             global_key="global_key1",
+        >>>             external_id="ex_id1",
+        >>>             attachments=[
+        >>>                 DataRowAttachmentSpec(type=AttachmentType.RAW_TEXT, name="att1", value="test1")
+        >>>             ],
+        >>>             metadata=[
+        >>>                 DataRowMetadataSpec(name="tag", value="tag value"),
+        >>>             ]
+        >>>         ),
+        >>>         # update existing data row by global key
+        >>>         DataRowSpec(
+        >>>             global_key="global_key1",
+        >>>             external_id="ex_id1_updated"
+        >>>         ),
+        >>>         # update global key of data row by existing global key
+        >>>         DataRowSpec(
+        >>>             key=DataRowGlobalKey("global_key1"),
+        >>>             global_key="global_key1_updated"
+        >>>         ),
+        >>>         # update data row by ID
+        >>>         DataRowSpec(
+        >>>             key=DataRowIdKey(dr.uid),
+        >>>             external_id="ex_id1_updated"
+        >>>         ),
+        >>>     ])
+        >>>     task.wait_till_done()
+        """
+        if len(specs) > MAX_DATAROW_PER_API_OPERATION:
+            raise MalformedQueryException(
+                f"Cannot upsert more than {MAX_DATAROW_PER_API_OPERATION} DataRows per function call."
+            )
+
+        class ManifestFile:
+
+            def __init__(self):
+                self.source = "SDK"
+                self.item_count = 0
+                self.chunk_uris: List[str] = []
+
+        def _convert_specs_to_upsert_items(_specs: List[DataRowSpec]):
+            _items: List[DataRowUpsertItem] = []
+            for spec in _specs:
+                # enforce current dataset's id for all specs
+                spec.dataset_id = self.uid
+                if spec.key:
+                    key = spec.key
+                elif spec.global_key:
+                    key = DataRowGlobalKey(spec.global_key)
+                else:
+                    key = DataRowAutoKey()
+                _items.append(DataRowUpsertItem(payload=spec, id=key))
+            return _items
+
+        items = _convert_specs_to_upsert_items(specs)
+        chunks = [
+            items[i:i + UPSERT_CHUNK_SIZE]
+            for i in range(0, len(items), UPSERT_CHUNK_SIZE)
+        ]
+        manifest = ManifestFile()
+        for chunk in chunks:
+            manifest.chunk_uris.append(
+                self._create_descriptor_file(chunk, is_upsert=True))
```

Comment:
I understand this is an optimization but did you consider uploading the chunks in parallel?